cluster-agent/controllers/argoproj.io/application_controller_test.go

```diff
@@ -0,0 +1,556 @@
+						Path:           "guestbook",
+						TargetRevision: "HEAD",
+						RepoURL:        "https://github.com/argoproj/argocd-example-apps.git",
+					},
+					Destination: appv1.ApplicationDestination{
+						Namespace: "guestbook",
+						Server:    "https://kubernetes.default.svc",
+					},
+					Project: "default",
+				},
+			}
+
+			reconciler = argo.ApplicationReconciler{
+				Client:        k8sClient,
+				Scheme:        scheme,
+				DB:            dbQueries,
+				TaskRetryLoop: sharedutil.NewTaskRetryLoop("application-reconciler"),
+			}
+		})
+
+		It("Create New Application CR in namespace and database and ApplicationState does not exist so it will create ApplicationState", func() {
+			// Close database connection.
+			defer dbQueries.CloseDatabase()
+
+			ctx = context.Background()
+
+			// Add databaseID label to applicationID.
+			databaseID := guestbookApp.Labels[dbID]
+			applicationDB := &db.Application{
+				Application_id:          databaseID,
+				Name:                    name,
+				Spec_field:              "{}",
+				Engine_instance_inst_id: gitopsEngineInstance.Gitopsengineinstance_id,
+				Managed_environment_id:  managedEnvironment.Managedenvironment_id,
+			}
+
+			// Create a new ArgoCD Application.
+			err = reconciler.Create(ctx, guestbookApp)
+			Expect(err).To(BeNil())
+
+			// Create Application in Database.
+			err = reconciler.DB.CreateApplication(ctx, applicationDB)
+			Expect(err).To(BeNil())
+
+			// Call reconcile function.
+			result, err := reconciler.Reconcile(ctx, newRequest(namespace, name))
+			Expect(err).To(BeNil())
+			Expect(result).ToNot(BeNil())
+
+		})
+
+		It("ApplicationState already exists, so just update it", func() {
+			// Close database connection.
+			defer dbQueries.CloseDatabase()
+
+			ctx = context.Background()
+
+			// Add databaseID label to applicationID.
+			databaseID := guestbookApp.Labels[dbID]
+			applicationDB := &db.Application{
+				Application_id:          databaseID,
+				Name:                    name,
+				Spec_field:              "{}",
+				Engine_instance_inst_id: gitopsEngineInstance.Gitopsengineinstance_id,
+				Managed_environment_id:  managedEnvironment.Managedenvironment_id,
+			}
+
+			applicationState := &db.ApplicationState{
+				Applicationstate_application_id: applicationDB.Application_id,
+				Health:                          "Healthy",
+				Sync_Status:                     "Synced",
+			}
+
+			// Create a new ArgoCD Application.
+			err = reconciler.Create(ctx, guestbookApp)
+			Expect(err).To(BeNil())
+
+			// Create Application in Database.
+			err = reconciler.DB.CreateApplication(ctx, applicationDB)
+			Expect(err).To(BeNil())
+
+			// Create ApplicationState in Database.
+			err = reconciler.DB.CreateApplicationState(ctx, applicationState)
+			Expect(err).To(BeNil())
+
+			// Call reconcile function.
+			result, err := reconciler.Reconcile(ctx, newRequest(namespace, name))
+			Expect(err).To(BeNil())
+			Expect(result).ToNot(BeNil())
+
```

Comment:
After `Reconcile` is called, you should verify that the `ApplicationState` row in the database is now equal to the `.status.health` and `.status.syncStatus` set in your `Application` resource.

Here we want to ensure the the Reconcile call will always update the `ApplicationState` row to be consistent with the Argo CD Application sync status and health, replacing the existing values in `ApplicationState` for these fields.